File: 백준/Gold/14002. 가장 긴 증가하는 부분 수열 4/가장 긴 증가하는 부분 수열 4.py
n = int(input())
lis = list(map(int,input().split()))

dp = [1]*n

for i in range(n):
  for j in range(0,i):
    if lis[j] < lis[i]: dp[i] = max(dp[j]+1,dp[i])

answer = []
order = max(dp)
for i in range(n-1,-1,-1):
  if dp[i] == order:
    answer.append(lis[i])
    order-=1



print(max(dp))
print(*answer[::-1])

File: 백준/Gold/14002. 가장 긴 증가하는 부분 수열 4/가장 긴 증가하는 부분 수열 4.js
const fs = require('fs')
const input = fs.readFileSync('/dev/stdin','utf8').trim().split('\n')

const N = Number(input.shift())
const values = input.shift().trim().split(' ').map(Number)

const dp = Array(N).fill(1)

for (let i = 0;i<N;i++){
    for (let j = 0;j<i;j++){
        if (values[i] > values[j]){
            dp[i] = Math.max(dp[i],dp[j]+1)
        }
    }
}

const answer = []
let order = Math.max(...dp)
for (let i = N-1;i>-1;i--){
    if (dp[i] === order){
        answer.push(values[i])        
        order--
    }
}

console.log(answer.length)
console.log(...answer.reverse())
